### PR TITLE
ExecutableWorld,  designed to also work with BatchWorld

### PR DESCRIPTION
**Pull request by jaseweston**
A world where messages from agents can be interpreted as _actions_ in the world which result in changes in the environment (are executed). Hence a grounded simulation can be implemented rather than just dialogue between agents.

### CODE REVIEW

File: parlai/core/worlds.py [hunk 1]
```diff
@@ -315,6 +315,53 @@ def shutdown(self):
...
+        """
+        pass
+
+    def observe(self, agent, act):
+        """An executable world class should implement this function, otherwise
+        the observations for each agent are just the messages from other agents
+        and not confitioned on the world at all (and it is thus the same as
+        MultiAgentDialogWorld). """
+        if agent.id == act['id']:
+            return None
```
Comment (alexholdenmiller) [inline parlai/core/worlds.py]: None or {}?

File: parlai/core/worlds.py [hunk 2]
```diff
@@ -315,6 +315,53 @@ def shutdown(self):
...
+        acts = self.acts
+        for index, agent in enumerate(self.agents):
+            # The agent acts.
+            acts[index] = agent.act()
+            # We execute this action in the world. 
+            self.execute(agent, acts[index])
+            # All agents (might) observe the results.
+            for other_agent in self.agents:
+                obs = self.observe(other_agent, acts[index])
+                if obs is not None:
```
Comment (alexholdenmiller) [inline parlai/core/worlds.py]: `if obs` instead of you change `None` to `{}`, but either way is fine with me

File: parlai/core/worlds.py [hunk 3]
```diff
@@ -481,11 +528,20 @@ def __next__(self):
...
         batch_observations = []
         for i, w in enumerate(self.worlds):
             agents = w.get_agents()
-            observation = agents[index].observe(validate(batch_actions[i]))
+            observation = None
+            if hasattr(w, 'observe'):
+                # The world has its own observe function, which the action
+                # first goes through (agents receive messages via the world,
+                # not from each other).
+                observation = w.observe(agents[index], validate(batch_actions[i]))
```
Comment (alexholdenmiller) [inline parlai/core/worlds.py]: does this have a chance of returning `None`?

File: parlai/core/worlds.py [hunk 4]
```diff
@@ -523,11 +579,17 @@ def parley(self):
...
+                if hasattr(w, 'execute'):
+                    w.execute(w.agents[i], batch_act[i])
+            # All agents (might) observe the results.
             for other_index in range(num_agents):
-                if index != other_index:
-                    batch_observations[other_index] = (
-                        self.batch_observe(other_index, batch_act))
+                obs = self.batch_observe(other_index, batch_act, index)
+                if obs is not None:
+                    batch_observations[other_index] = obs
```
Comment (alexholdenmiller) [inline parlai/core/worlds.py]: I wonder if you need to override this always to make sure that you don't accidentally view a stale message? maybe better to fill it with `{}`

**Comment by jaseweston** [comment]
I just didn't want to break existing code that doesn't set anything. I think all models should be able to cope with an empty observation though, ideally. So, in a batch there could be empty ones. But right now we don't allow for that.. should I change it?

**Comment by alexholdenmiller** [comment]
we can follow up later with the empty table changes

**Pull request merged**